README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# MiceMf

## Random-forest-based Single Imputation adapted for MICE

## Installation for development version from GitHub

Currently the `MiceMf` package is only available as source files on
GitHub, any interested user can install and test the `MiceMf` package by
running the following code in R:

``` r
if(!"remotes" %in% installed.packages()) install.packages("remotes")
remotes::install_github("shangzhi-hong/MiceMf")
```

## Example

### Using warpper function

``` r
library(MiceMf)
# Perform imputation
data("airquality")
# Get the imputed data
resCmp <- MiceMf(airquality)
resCmpNorm <- MiceMfNorm(airquality)
```

### Using mice

``` r
library(MiceMf)
# Perform imputation
data("airquality")
impObj <- mice(airquality, method = "mf", m = 1, printFlag = FALSE)
impObjNorm <- mice(airquality, method = "mf.norm", m = 1, printFlag = FALSE)
# Get the complete data
resCmp <- complete(impObj, action = 1L)
resCmpNorm <- complete(impObjNorm, action = 1L)
```

### MissForest-flavored imputation using MICE

``` r
library(MiceMf)
# Perform imputation
data("airquality")
impObj <- missForestMice(airquality)
```


NEWS.md
# MiceMf 1.0.0
* First public release on GitHub
* Unit tests ready


tests/testthat.R
library(testthat)
library(MiceMf)

test_check("MiceMf")


tests/testthat/test-Imp-MfMice.R
context("Imputation using caller")

# Set-up
NUM_OBS <- 50
testData <- data.frame(
    x1 = rnorm(NUM_OBS, mean = 2, sd = 1),
    x2 = rnorm(NUM_OBS, mean = 2, sd = 1),
    x3 = rnorm(NUM_OBS, mean = 2, sd = 1)
)

testData <- transform(testData,
                      y = 2 + 2 * x1 + 3 * x2 + 4 * x3 + rnorm(NUM_OBS))

amputedData <- ampute(
    data = testData,
    prop = 0.5,
    mech = "MCAR",
    bycases = TRUE,
    patterns = matrix(
        data = c(0, 0, 0, 1,
                 1, 0, 1, 1),
        ncol = 4,
        byrow = T
    )
)[["amp"]]

test_that("missForestMice works for continuous variables", {
    for (decreasing in c(TRUE, FALSE)) {
        for (maxiter in c(5, 10)) {
            impDf <- missForestMice(xmis = amputedData,
                                    decreasing = decreasing)

            expect_true(!anyNA(impDf))
        }
    }

})

test_that("MiceMf works for continuous variables", {
    for (decreasing in c(TRUE, FALSE)) {
        for (maxiter in c(5, 10)) {
            impDf <- MiceMf(data = amputedData,
                            decreasing = decreasing)
            expect_true(!anyNA(impDf))
        }
    }
})

test_that("MiceMfNorm works for continuous variables", {
    for (decreasing in c(TRUE, FALSE)) {
        for (maxiter in c(5, 10)) {
            impDf <- MiceMfNorm(data = amputedData,
                                decreasing = decreasing)
            expect_true(!anyNA(impDf))
        }
    }

})


tests/testthat/test-Imp-Catgorical.R
context("Imputation for categorical variables")

# Set-up
NUM_OBS <- 100
testData <- data.frame(
    x1 = rnorm(NUM_OBS, mean = 2, sd = 1),
    x2 = rnorm(NUM_OBS, mean = 2, sd = 1),
    x3 = rnorm(NUM_OBS, mean = 2, sd = 1)
)

testData <- transform(
    testData,
    y =  {
        yTemp <- 2 + 2 * x1 + 3 * x2 + 4 * x3 + rnorm(NUM_OBS)
        y <-
            cut(yTemp, breaks = quantile(
                    yTemp,
                    probs = seq(0, 1, by = 0.20),
                    na.rm = TRUE,
                    names = FALSE),
                include.lowest = TRUE,
                right = TRUE,
                labels = c("A", "B", "C", "D", "E")
            )
        y
    }
)

y <- testData[["y"]]
x <- subset(testData, select = -y)
ry <- sample(
    x = c(TRUE, FALSE),
    size = NUM_OBS,
    replace = TRUE,
    prob = c(0.5, 0.5))
wy <- !ry
num.trees.categorical <- 10

test_that("mf.categorical works for categorical variables", {
    for (pre.bootstrap in c(TRUE, FALSE)) {
        for (use.pred.prob.categorical in c(TRUE, FALSE)) {
            catImpOut1 <- mice.impute.mf.categorical(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical,
                num.threads = 1)
            catImpOut2 <- mice.impute.mf.categorical(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical,
                num.threads = 1)
            expect_true(all(c(is.factor(catImpOut1), is.factor(catImpOut2))))
            expect_true(length(catImpOut1) == sum(wy))
            expect_true(length(catImpOut2) == sum(wy))
            expect_true(!anyNA(catImpOut1) && !anyNA(catImpOut2))
            expect_true(!all(catImpOut1 == catImpOut2))
        }
    }

})

test_that("mf caller works for categorical variables", {
    for (pre.bootstrap in c(TRUE, FALSE)) {
        for (use.pred.prob.categorical in c(TRUE, FALSE)) {
            catImpOut1 <- mice.impute.mf(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            catImpOut2 <- mice.impute.mf(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            expect_true(all(c(is.factor(catImpOut1), is.factor(catImpOut2))))
            expect_true(length(catImpOut1) == sum(wy))
            expect_true(length(catImpOut2) == sum(wy))
            expect_true(!anyNA(catImpOut1) && !anyNA(catImpOut2))
            expect_true(!all(catImpOut1 == catImpOut2))
        }
    }
})

# Logical

testData <- transform(
    testData,
    y =  {
        yTemp <- 2 + 2 * x1 + 3 * x2 + 4 * x3 + rnorm(NUM_OBS)
        yQ50 <- quantile(yTemp, probs = 0.5, na.rm = TRUE, names = FALSE)
        y <- yTemp >= yQ50
        y
    }
)

y <- testData[["y"]]
x <- subset(testData, select = -y)

test_that("mf.categorical works for logical variables", {
    for (pre.bootstrap in c(TRUE, FALSE)) {
        for (use.pred.prob.categorical in c(TRUE, FALSE)) {
            catImpOut1 <- mice.impute.mf.categorical(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            catImpOut2 <- mice.impute.mf.categorical(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            expect_true(all(c(is.logical(catImpOut1), is.logical(catImpOut2))))
            expect_true(length(catImpOut1) == sum(wy))
            expect_true(length(catImpOut2) == sum(wy))
            expect_true(!anyNA(catImpOut1) && !anyNA(catImpOut2))
            expect_true(!all(catImpOut1 == catImpOut2))
        }
    }

})

test_that("mf caller works for logical variables", {
    for (pre.bootstrap in c(TRUE, FALSE)) {
        for (use.pred.prob.categorical in c(TRUE, FALSE)) {
            catImpOut1 <- mice.impute.mf(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            catImpOut2 <- mice.impute.mf(
                y = y,
                ry = ry,
                x = x,
                wy = wy,
                num.trees.categorical = num.trees.categorical,
                pre.bootstrap = pre.bootstrap,
                use.pred.prob.categorical = use.pred.prob.categorical)
            expect_true(all(c(is.logical(catImpOut1), is.logical(catImpOut2))))
            expect_true(length(catImpOut1) == sum(wy))
            expect_true(length(catImpOut2) == sum(wy))
            expect_true(!anyNA(catImpOut1) && !anyNA(catImpOut2))
            expect_true(!all(catImpOut1 == catImpOut2))
        }
    }
})


R/mice.impute.mf.R
#' Single imputation using random forest algorithm
#'
#' @param y Vector to be imputed
#'
#' @param ry Logical vector of length \code{length(y)} indicating the
#' the subset \code{y[ry]} of elements in \code{y} to which the imputation
#' model is fitted. The \code{ry} generally distinguishes the observed
#' (\code{TRUE}) and missing values (\code{FALSE}) in \code{y}.
#'
#' @param x Numeric design matrix with \code{length(y)} rows with predictors for
#' \code{y}. Matrix \code{x} may have no missing values.
#'
#' @param wy Logical vector of length \code{length(y)}. A \code{TRUE} value
#' indicates locations in \code{y} for which imputations are created.
#'
#' @param num.trees Number of trees to build, default to \code{10}, overrides
#' \code{num.trees.continuous} and \code{num.trees.categorical}
#'
#' @param num.trees.continuous Number of trees to build for imputing continuous
#' variables, default to \code{10}
#'
#' @param num.trees.categorical Number of trees to build for imputing
#' categorical variables, default to \code{10}
#'
#' @param use.pred.prob.categorical Logical indicator, \code{TRUE} for
#' imputation based on probabilities of votes, \code{FALSE} for imputation
#' based on majority votes, default to \code{TRUE}
#'
#' @param pre.bootstrap Perform bootstrap prior to imputation to get 'proper'
#' imputation, i.e. accommodating sampling variation in estimating population
#' regression parameters (see Shah et al. 2014)
#'
#' @param ... Other arguments to pass down
#'
#' @return Vector with imputed data, same type as \code{y}, and of length
#' \code{sum(wy)}
#'
#' @author Shangzhi Hong, Henry S. Lynn*
#'
#' @references
#'
#' Stekhoven, Daniel J., and Peter Bühlmann.
#' "MissForest—non-parametric missing value imputation for mixed-type data."
#' Bioinformatics 28.1 (2012): 112-118.
#'
#' Shah, Anoop D., et al. "Comparison of random forest and parametric
#' imputation models for imputing missing data using MICE: a CALIBER study."
#' American journal of epidemiology 179.6 (2014): 764-774.
#'
#' @examples
#' imp <- mice(nhanes, method = "mf")
#'
#' @export
mice.impute.mf <- function(
    y,
    ry,
    x,
    wy = NULL,
    num.trees = 10,
    num.trees.continuous = NULL,
    num.trees.categorical = NULL,
    use.pred.prob.categorical = FALSE,
    pre.bootstrap = FALSE,
    ...
) {
    if (is.numeric(y)) {
        if (is.null(num.trees.continuous)) num.trees.continuous <- num.trees
        rfImpVec <- mice.impute.mf.continuous(
            y = y,
            ry = ry,
            x = x,
            wy = wy,
            num.trees.continuous = num.trees.continuous,
            normal.error.continuous = FALSE,
            pre.bootstrap = pre.bootstrap,
            ...
        )
    } else {
        if (is.null(num.trees.categorical)) num.trees.categorical <- num.trees
        rfImpVec <- mice.impute.mf.categorical(
            y = y,
            ry = ry,
            x = x,
            wy = wy,
            num.trees.categorical = num.trees.categorical,
            use.pred.prob.categorical = use.pred.prob.categorical,
            pre.bootstrap = pre.bootstrap,
            ...
        )
    }
    return(rfImpVec)
}


R/missForestMice.R
#' MissForest-flavored missing data imputation under MICE framework
#'
#' @param xmis Data set containing missing values
#'
#' @param maxiter Number of imputation iterations
#'
#' @param ntree Number of trees to build
#'
#' @param mtry Number of variables randomly sampled at each split
#'
#' @param decreasing If \code{FALSE} then the variables are sorted according
#' increasing amount of missing cells.
#'
#' @param ... Other arguments to pass down
#'
#' @return Imputed complete dataset
#'
#' @author Shangzhi Hong, Henry S. Lynn*
#'
#' @references
#'
#' Stekhoven, Daniel J., and Peter Bühlmann.
#' "MissForest—non-parametric missing value imputation for mixed-type data."
#' Bioinformatics 28.1 (2012): 112-118.
#'
#' Shah, Anoop D., et al. "Comparison of random forest and parametric
#' imputation models for imputing missing data using MICE: a CALIBER study."
#' American journal of epidemiology 179.6 (2014): 764-774.
#'
#' @examples
#' imp <- missForestMice(nhanes)
#'
#' @export
missForestMice <- function(xmis,
                           maxiter = 5,
                           ntree = 10,
                           mtry = floor(sqrt(ncol(xmis))),
                           decreasing = FALSE,
                           ...) {
    # Set-up missForest-flavored initial value
    xInit <- xmis
    xAttrib <- lapply(xmis, attributes)
    numVar <- ncol(xmis)
    for (varIdx in 1:numVar) {
        if (is.null(xAttrib[[varIdx]])) {
            xInit[is.na(xmis[, varIdx]), varIdx] <-
                mean(xmis[, varIdx], na.rm = TRUE)
        } else {
            max.level <- max(table(xInit[, varIdx]))
            class.assign <-
                sample(names(which(max.level == summary(xInit[, varIdx]))), 1)
            if (class.assign != "NA's") {
                xInit[is.na(xmis[, varIdx]), varIdx] <- class.assign
            } else {
                while (class.assign == "NA's") {
                    class.assign <- sample(names(which(
                        max.level ==
                            summary(xInit[, varIdx])
                    )), 1)
                }
                xInit[is.na(xmis[, varIdx]), varIdx] <- class.assign
            }
        }
    }

    miceImp <- mice(
        data = xmis,
        m = 1,
        method = "mf",
        maxit = maxiter,
        visitSequence = ifelse(decreasing, "revmonotone", "monotone"),
        printFlag = FALSE,
        data.init = xInit,
        num.trees = ntree,
        maxcor = 1.0,
        ...
    )
    miceCompData <- complete(data = miceImp, action = 1L)
    return(miceCompData)
}


R/mice.impute.mf.continuous.R
#' Single imputation for continuous variable based random forest
#'
#' @param y Vector to be imputed
#'
#' @param ry Logical vector of length \code{length(y)} indicating the
#' the subset \code{y[ry]} of elements in \code{y} to which the imputation
#' model is fitted. The \code{ry} generally distinguishes the observed
#' (\code{TRUE}) and missing values (\code{FALSE}) in \code{y}.
#'
#' @param x Numeric design matrix with \code{length(y)} rows with predictors for
#' \code{y}. Matrix \code{x} may have no missing values.
#'
#' @param wy Logical vector of length \code{length(y)}. A \code{TRUE} value
#' indicates locations in \code{y} for which imputations are created.
#'
#' @param num.trees.continuous Number of trees to build for imputing continuous
#' variables, default to \code{10}
#'
#' @param pre.bootstrap Perform bootstrap prior to imputation to get 'proper'
#' imputation, i.e. accommodating sampling variation in estimating population
#' regression parameters (see Shah et al. 2014).
#'
#' @param normal.error.continuous Logical indicator, \code{TRUE} for
#' assuming normal distribution for error,
#' the variance equals to overall out-of-bag prediction error,
#' i.e. mean squared error (see Shah et al. 2014).
#'
#' @param ... Other arguments to pass down
#'
#' @return Vector with imputed data, same type as \code{y}, and of length
#' \code{sum(wy)}
#'
#' @author Shangzhi Hong, Henry S. Lynn*
#'
#' @export
mice.impute.mf.continuous <- function(
    y,
    ry,
    x,
    wy = NULL,
    num.trees.continuous = 10,
    pre.bootstrap = FALSE,
    normal.error.continuous = FALSE,
    ...
    ) {
    if (is.null(wy)) wy <- !ry
    yMisNum <- sum(wy)
    if (isTRUE(pre.bootstrap)) {
        bootIdx <- sample(sum(ry), replace = TRUE)
        yObs <- y[ry][bootIdx]
        xObs <- x[ry, , drop = FALSE][bootIdx, , drop = FALSE]
    } else {
        yObs <- y[ry]
        xObs <- x[ry, , drop = FALSE]
    }
    xMis <- x[wy, , drop = FALSE]
    rangerObj <- ranger(
        x = xObs,
        y = yObs,
        oob.error = TRUE,
        num.trees = num.trees.continuous)
    misPredVal <- predictions(predict(rangerObj, xMis))
    if (normal.error.continuous) {
        # Use overall out-of-bag prediction error
        impVal <- rnorm(length(misPredVal),
                        mean = misPredVal,
                        sd = sqrt(rangerObj[["prediction.error"]]))
    } else {
        impVal <- misPredVal
    }
    return(impVal)
}


R/mice.impute.mf.categorical.R
#' Single imputation for categorical variable based random forest
#'
#' @param y Vector to be imputed
#'
#' @param ry Logical vector of length \code{length(y)} indicating the
#' the subset \code{y[ry]} of elements in \code{y} to which the imputation
#' model is fitted. The \code{ry} generally distinguishes the observed
#' (\code{TRUE}) and missing values (\code{FALSE}) in \code{y}.
#'
#' @param x Numeric design matrix with \code{length(y)} rows with predictors for
#' \code{y}. Matrix \code{x} may have no missing values.
#'
#' @param wy Logical vector of length \code{length(y)}. A \code{TRUE} value
#' indicates locations in \code{y} for which imputations are created.
#'
#' @param num.trees.categorical Number of trees to build for imputing
#' categorical variables, default to \code{10}.
#'
#' @param use.pred.prob.categorical Logical indicator, \code{TRUE} for
#' imputation based on probabilities of votes, \code{FALSE} for imputation
#' based on majority votes, default to \code{TRUE}
#'
#' @param pre.bootstrap Perform bootstrap prior to imputation to get 'proper'
#' imputation, i.e. accommodating sampling variation in estimating population
#' regression parameters (see Shah et al. 2014)
#'
#' @param ... Other arguments to pass down
#'
#' @return Vector with imputed data, same type as \code{y}, and of length
#' \code{sum(wy)}
#'
#' @author Shangzhi Hong, Henry S. Lynn*
#'
#' @export
mice.impute.mf.categorical <- function(
    y,
    ry,
    x,
    wy = NULL,
    num.trees.categorical = 10,
    use.pred.prob.categorical = FALSE,
    pre.bootstrap = FALSE,
    ...){
    if (is.null(wy)) wy <- !ry
    yMisNum <- sum(wy)
    if (is.logical(y)) {
        yIsLogical <- TRUE
        y <- as.factor(y)
    } else {
        yIsLogical <- FALSE
    }
    if (isTRUE(pre.bootstrap)) {
        bootIdx <- sample(sum(ry), replace = TRUE)
        yObs <- y[ry][bootIdx]
        xObs <- x[ry, , drop = FALSE][bootIdx, , drop = FALSE]
    } else {
        yObs <- y[ry]
        xObs <- x[ry, , drop = FALSE]
    }
    # Short-cut
    yobsLvDrop <- droplevels(yObs)
    yObsLvNum <- nlevels(yobsLvDrop)
    if (yObsLvNum == 1)
        return(rep_len(yObs, length.out = yMisNum))

    xMis <- x[wy, , drop = FALSE]
    if (isTRUE(use.pred.prob.categorical)) {
        # Construct predictions based on vote probs averaged over nodes.
        # Suppress warnings like:
        # "Dropped unused factor level(s) in dependent variable"
        rangerObjProb <- suppressWarnings(ranger(
            x = xObs,
            y = yObs,
            probability = TRUE,
            oob.error = FALSE,
            num.trees = num.trees.categorical))
        misPredMat <- predictions(predict(rangerObjProb, xMis))
        yLevels <- colnames(misPredMat)
        impValChar <- apply(
            X = misPredMat,
            MARGIN = 1,
            FUN = function(voteProb) {
                sample(x = yLevels, size = 1, prob = voteProb)
            })
        impVal <- factor(x = impValChar, levels = levels(y))
    } else {
        # Construct predictions based on major votes, less variant
        rangerObj <- ranger(
            x = xObs,
            y = yObs,
            oob.error = FALSE,
            num.trees = num.trees.categorical)
        impVal <- predictions(predict(rangerObj, xMis))
    }
    if (yIsLogical) impVal <- as.logical(impVal == "TRUE")
    return(impVal)
}


R/mice.impute.mfnorm.R
#' Single imputation using random forest algorithm with normality assumption
#' for prediction error
#'
#' @param y Vector to be imputed
#'
#' @param ry Logical vector of length \code{length(y)} indicating the
#' the subset \code{y[ry]} of elements in \code{y} to which the imputation
#' model is fitted. The \code{ry} generally distinguishes the observed
#' (\code{TRUE}) and missing values (\code{FALSE}) in \code{y}.
#'
#' @param x Numeric design matrix with \code{length(y)} rows with predictors for
#' \code{y}. Matrix \code{x} may have no missing values.
#'
#' @param wy Logical vector of length \code{length(y)}. A \code{TRUE} value
#' indicates locations in \code{y} for which imputations are created.
#'
#' @param num.trees Number of trees to build, default to \code{10}, overrides
#' \code{num.trees.continuous} and \code{num.trees.categorical}
#'
#' @param num.trees.continuous Number of trees to build for imputing continuous
#' variables, default to \code{10}
#'
#' @param num.trees.categorical Number of trees to build for imputing
#' categorical variables, default to \code{10}
#'
#' @param use.pred.prob.categorical Logical indicator, \code{TRUE} for
#' imputation based on probabilities of votes, \code{FALSE} for imputation
#' based on majority votes, default to \code{TRUE}
#'
#' @param pre.bootstrap Perform bootstrap prior to imputation to get 'proper'
#' imputation, i.e. accommodating sampling variation in estimating population
#' regression parameters (see Shah et al. 2014)
#'
#' @param ... Other arguments to pass down
#'
#' @return Vector with imputed data, same type as \code{y}, and of length
#' \code{sum(wy)}
#'
#' @author Shangzhi Hong, Henry S. Lynn*
#'
#' @references
#'
#' Stekhoven, Daniel J., and Peter Bühlmann.
#' "MissForest—non-parametric missing value imputation for mixed-type data."
#' Bioinformatics 28.1 (2012): 112-118.
#'
#' Shah, Anoop D., et al. "Comparison of random forest and parametric
#' imputation models for imputing missing data using MICE: a CALIBER study."
#' American journal of epidemiology 179.6 (2014): 764-774.
#'
#' @examples
#' imp <- mice(nhanes, method = "mf.norm")
#'
#' @export
mice.impute.mf.norm <- function(
    y,
    ry,
    x,
    wy = NULL,
    num.trees = 10,
    num.trees.continuous = NULL,
    num.trees.categorical = NULL,
    use.pred.prob.categorical = FALSE,
    pre.bootstrap = FALSE,
    ...
) {
    if (is.numeric(y)) {
        if (is.null(num.trees.continuous)) num.trees.continuous <- num.trees
        rfImpVec <- mice.impute.mf.continuous(
            y = y,
            ry = ry,
            x = x,
            wy = wy,
            num.trees.continuous = num.trees.continuous,
            normal.error.continuous = TRUE,
            pre.bootstrap = pre.bootstrap,
            ...
        )
    } else {
        if (is.null(num.trees.categorical)) num.trees.categorical <- num.trees
        rfImpVec <- mice.impute.mf.categorical(
            y = y,
            ry = ry,
            x = x,
            wy = wy,
            num.trees.categorical = num.trees.categorical,
            use.pred.prob.categorical = use.pred.prob.categorical,
            pre.bootstrap = pre.bootstrap,
            ...
        )
    }
    return(rfImpVec)
}
